src/server/ua_subscription_alarms_conditions.c

```diff
@@ -0,0 +1,2575 @@
+                                    UA_TwoStateVariableCallbackType callbackType) {
+    switch(callbackType) {
+        case UA_ENTERING_ENABLEDSTATE:
+            if(conditionEntry->specificCallbacksData.enteringEnabledStateCallback != NULL) {
+              return conditionEntry->specificCallbacksData.enteringEnabledStateCallback(server, branch);
+            }
+            return UA_STATUSCODE_GOOD;//TODO log warning when the callback wasn't set
+
+        case UA_ENTERING_ACKEDSTATE:
+            if(conditionEntry->specificCallbacksData.enteringAckedStateCallback != NULL) {
+              *removeBranch = conditionEntry->specificCallbacksData.ackedRemoveBranch;
+              return conditionEntry->specificCallbacksData.enteringAckedStateCallback(server, branch);
+            }
+            return UA_STATUSCODE_GOOD;
+
+        case UA_ENTERING_CONFIRMEDSTATE:
+            if(conditionEntry->specificCallbacksData.enteringConfirmedStateCallback != NULL) {
+              *removeBranch = conditionEntry->specificCallbacksData.confirmedRemoveBranch;
+              return conditionEntry->specificCallbacksData.enteringConfirmedStateCallback(server, branch);
+            }
+            return UA_STATUSCODE_GOOD;
+
+        case UA_ENTERING_ACTIVESTATE:
+            if(conditionEntry->specificCallbacksData.enteringActiveStateCallback != NULL) {
+              return conditionEntry->specificCallbacksData.enteringActiveStateCallback(server, branch);
+            }
+            return UA_STATUSCODE_GOOD;
+
+        default:
+            return UA_STATUSCODE_BADNOTFOUND;
+      }
+}
+
+static UA_StatusCode
+callConditionTwoStateVariableCallback(UA_Server *server,
+                                      const UA_NodeId *condition,
+                                      const UA_NodeId *conditionSource,
+                                      UA_Boolean *removeBranch,
+                                      UA_TwoStateVariableCallbackType callbackType) {
+    /* Get ConditionSource Entry */
+    UA_ConditionSource_nodeListElement *conditionSourceEntryTmp;
+    LIST_FOREACH(conditionSourceEntryTmp, &server->headConditionSource, listEntry) {
+        if(!UA_NodeId_equal(&conditionSourceEntryTmp->conditionSourceId, conditionSource))
+            continue;
+        /* Get Condition Entry */
+        UA_Condition_nodeListElement *conditionEntryTmp;
+        LIST_FOREACH(conditionEntryTmp, &conditionSourceEntryTmp->conditionHead, listEntry) {
+            if(UA_NodeId_equal(&conditionEntryTmp->conditionId, condition)) {
+                return getonditionTwoStateVariableCallback(server, condition, conditionEntryTmp, removeBranch, callbackType);
+            }
+            else {
+                UA_ConditionBranch_nodeListElement *conditionBranchEntryTmp;
+                LIST_FOREACH(conditionBranchEntryTmp, &conditionEntryTmp->conditionBranchHead, listEntry) {
+                    if(UA_NodeId_equal(conditionBranchEntryTmp->conditionBranchId, condition))
```

Comment:
I believe that checking for a null pointer is needed here. I'm getting a segfault here since the first branch (trunk) has its id == NULL.
```suggestion
                    if(conditionBranchEntryTmp->conditionBranchId != NULL &&
                       UA_NodeId_equal(conditionBranchEntryTmp->conditionBranchId, condition))
```

Reply:
you are right:)
The missing Null pointer check was because the previous implementation of `UA_NodeId_equal` did that check.